## Investigasi sampel data titanic berikut dengan cara :
1. Cek secara head, tail, sample, info lalu observasi apa yang bisa anda peroleh ?
2. Lakukan Statistical Summary dengan mengekstrak informasi yang didapat dari observasi anda ?
3. Cek apakah ada duplikat dan bagaimana handlenya ?
4. Cek apakah ada missing value, berapa persentasenya jika ada, dan bagaimana cara handlenya ?

# Import Libraries

In [ ]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", None)

# Load Data

In [ ]:
# import data
df = pd.read_excel('titanic.xlsx')
data = df.copy()

#  Preliminary

In [ ]:
data.head()

survived                                             name     sex      age
0         1                    Allen, Miss. Elisabeth Walton  female  29.0000
1         1                   Allison, Master. Hudson Trevor    male   0.9167
2         0                     Allison, Miss. Helen Loraine  female   2.0000
3         0             Allison, Mr. Hudson Joshua Creighton    male  30.0000
4         0  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  female  25.0000

In [ ]:
data.tail()

survived                                     name     sex   age
495         1  Mallet, Mrs. Albert (Antoinette Magnin)  female  24.0
496         0        Mangiavacchi, Mr. Serafino Emilio    male   NaN
497         0               Matthews, Mr. William John    male  30.0
498         0                Maybery, Mr. Frank Hubert    male  40.0
499         0                McCrae, Mr. Arthur Gordon    male  32.0

In [ ]:
data.sample(5)

survived                                               name     sex  \
452         0                                  Hold, Mr. Stephen    male   
99          1  Duff Gordon, Lady. (Lucille Christiana Sutherl...  female   
438         1                                Herman, Miss. Alice  female   
277         1     Spencer, Mrs. William Augustus (Marie Eugenie)  female   
1           1                     Allison, Master. Hudson Trevor    male   

         age  
452  44.0000  
99   48.0000  
438  24.0000  
277      NaN  
1     0.9167

Observation:
1. Columns `name` and `sex` are categorical, columns `survived` and `age` are numeric.
2. `sex` column seems to contain two distinct values (male OR female), but will confirm later
3. `survived` is apparently also binary (0,1)
4. No obvious defect on the data (column name vs its entries), all looks good



In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 500 entries, 0 to 499
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   survived  500 non-null    int64  
 1   name      500 non-null    object 
 2   sex       500 non-null    object 
 3   age       451 non-null    float64
dtypes: float64(1), int64(1), object(2)
memory usage: 15.8+ KB


Observation:
1. Data contains 4 columns with 500 rows
2. Only the age column has missing values, it will be handled later
3. All dtypes seem OK (appropriate), given the corresponding column name

# Statistical Summary

In [ ]:
data.columns

Index(['survived', 'name', 'sex', 'age'], dtype='object')

In [ ]:
# group column names based on type
categoricals = ['name', 'sex']

numericals = ['survived', 'age']

In [ ]:
data[numericals].describe()

survived         age
count  500.000000  451.000000
mean     0.540000   35.917775
std      0.498897   14.766454
min      0.000000    0.666700
25%      0.000000   24.000000
50%      1.000000   35.000000
75%      1.000000   47.000000
max      1.000000   80.000000

Observations:

*   Overall, the minimum and maximum values make sense for each column
*   Mean ~50% (Median) in the `age` column indicates a somewhat symmetrical distribution
*   `survived` column is boolean/binary column since the value is 0 or 1, no need to conclude its simmetricity. Only need to check balance level








In [ ]:
data[categoricals].describe()

name   sex
count                              500   500
unique                             499     2
top     Eustis, Miss. Elizabeth Mussey  male
freq                                 2   288

Observations:
* `sex` has 2 unique values, 'male' OR 'female'
* The most passengers' gender is male (288 rows), the rest are female.
* in the `name` column there are 2 data that have the same name (Eustis, Miss. Elizabeth Mussey), it is necessary to check further whether the data is duplicate data.

In [ ]:
for col in categoricals:
  print(f"Value counts of {col} column")
  print(data[col].value_counts(), '\n')

Value counts of name column
name
Eustis, Miss. Elizabeth Mussey                                                        2
Becker, Miss. Ruth Elizabeth                                                          1
Becker, Miss. Marion Louise                                                           1
Becker, Master. Richard F                                                             1
Beauchamp, Mr. Henry James                                                            1
Beane, Mrs. Edward (Ethel Clarke)                                                     1
Beane, Mr. Edward                                                                     1
Bateman, Rev. Robert James                                                            1
Banfield, Mr. Frederick James                                                         1
Ball, Mrs. (Ada E Hall)                                                               1
Baimbrigge, Mr. Charles Robert                                                        1

In [ ]:
for col in numericals:
  print(f"==== {col} ====")
  print(data[col].value_counts(), '\n')

==== survived ====
survived
1    270
0    230
Name: count, dtype: int64 

==== age ====
age
24.0000    23
30.0000    20
36.0000    19
18.0000    14
42.0000    14
45.0000    14
35.0000    14
22.0000    12
28.0000    12
23.0000    11
21.0000    11
31.0000    11
48.0000    11
25.0000    11
29.0000    10
54.0000    10
50.0000    10
34.0000    10
47.0000    10
39.0000     9
27.0000     9
33.0000     8
49.0000     8
38.0000     8
19.0000     8
40.0000     8
55.0000     8
32.0000     7
60.0000     7
17.0000     7
26.0000     7
44.0000     6
52.0000     6
37.0000     6
58.0000     6
64.0000     5
46.0000     5
57.0000     5
51.0000     5
16.0000     4
53.0000     4
56.0000     4
43.0000     4
62.0000     4
61.0000     4
1.0000      3
8.0000      3
63.0000     3
41.0000     3
15.0000     2
71.0000     2
20.0000     2
65.0000     2
4.0000      2
6.0000      2
2.0000      1
0.9167      1
14.0000     1
11.0000     1
76.0000     1
59.0000     1
80.0000     1
28.5000     1
32.5000     1
70.0000     

# Cleaning Data

## 1. Duplicate Handling

In [ ]:
len(data)

500

In [ ]:
len(data.drop_duplicates())

499

In [ ]:
len(data.drop_duplicates()) / len(data)
#if the output of the code in this cell is not 1 then there are duplicates

0.998

In [ ]:
# Step 1: Get duplicate rows (including the original one)
duplicates = data[data.duplicated(keep=False)]

In [ ]:
duplicates

survived                            name     sex   age
104         1  Eustis, Miss. Elizabeth Mussey  female  54.0
349         1  Eustis, Miss. Elizabeth Mussey  female  54.0

In [ ]:
duplicates.groupby(list(data.columns)).size().reset_index(name = 'duplicate_count')

survived                            name     sex   age  duplicate_count
0         1  Eustis, Miss. Elizabeth Mussey  female  54.0                2

it is verified a row that has the same name (Eustis, Miss. Elizabeth Mussey) is duplicate

In [ ]:
#Handling Drop duplicate
data = data.drop_duplicates()

In [ ]:
len(data.drop_duplicates()) / len(data)
#if the output of the code in this cell is not 1 then there are duplicates

1.0

### Has handled duplicate drops

## 2. Missing value handling
### Identifying Missing value

In [ ]:
data.isna().sum()

survived     0
name         0
sex          0
age         49
dtype: int64

missing values only in `age` column

In [ ]:
for column in data.columns:
    print(f"============= {column} =================")
    display(data[column].value_counts())
    print()

============= survived =================


survived
1    269
0    230
Name: count, dtype: int64


============= name =================


name
McCrae, Mr. Arthur Gordon                                                             1
Allen, Miss. Elisabeth Walton                                                         1
Allison, Master. Hudson Trevor                                                        1
Lehmann, Miss. Bertha                                                                 1
Laroche, Mrs. Joseph (Juliette Marie Louise Lafargue)                                 1
Laroche, Mr. Joseph Philippe Lemercier                                                1
Laroche, Miss. Simonne Marie Anne Andree                                              1
Laroche, Miss. Louise                                                                 1
Lamb, Mr. John Joseph                                                                 1
Lahtinen, Rev. William                                                                1
Lahtinen, Mrs. William (Anna Sylfven)                                                 1
Kvillner, Mr. Johan Henrik Johannesson                                                1
Knight, Mr. Robert J                                                                  1
Kirkland, Rev. Charles Leonard                                                        1
Kelly, Mrs. Florence "Fannie"                                                         1
Keane, Mr. Daniel                                                                     1
Keane, Miss. Nora A                                                                   1
Karnes, Mrs. J Frank (Claire Bennett)                                                 1
Kantor, Mrs. Sinai (Miriam Sternin)                                                   1
Borebank, Mr. John James                                                              1
Bonnell, Miss. Elizabeth                                                              1
Bonnell, Miss. Caroline                                                               1
Blank, Mr. Henry                                                                      1
Blackwell, Mr. Stephen Weart                                                          1
Bjornstrom-Steffansson, Mr. Mauritz Hakan                                             1
Bissette, Miss. Amelia                                                                1
Bishop, Mrs. Dickinson H (Helen Walton)                                               1
Bishop, Mr. Dickinson H                                                               1
Birnbaum, Mr. Jakob                                                                   1
Bird, Miss. Ellen                                                                     1
Bidois, Miss. Rosalie                                                                 1
Behr, Mr. Karl Howell                                                                 1
Beckwith, Mrs. Richard Leonard (Sallie Monypeny)                                      1
Beckwith, Mr. Richard Leonard                                                         1
Beattie, Mr. Thomson                                                                  1
Cardeza, Mrs. James Warburton Martinez (Charlotte Wardle Drake)                       1
Cardeza, Mr. Thomas Drake Martinez                                                    1
Candee, Mrs. Edward (Helen Churchill Hungerford)                                      1
Calderhead, Mr. Edward Pennington                                                     1
Cairns, Mr. Alexander                                                                 1
Butt, Major. Archibald Willingham                                                     1
Burns, Miss. Elizabeth Margaret                                                       1
Bucknell, Mrs. William Robert (Emma Eliza Ward)                                       1
Brown, Mrs. John Murray (Caroline Lane Lamson)                                        1
Brown, Mrs. James Joseph (Margaret Tobin)                                             1
Brewe, Dr. Arthur Jackson          


============= sex =================


sex
male      288
female    211
Name: count, dtype: int64


============= age =================


age
24.0000    23
30.0000    20
36.0000    19
18.0000    14
42.0000    14
45.0000    14
35.0000    14
22.0000    12
28.0000    12
23.0000    11
21.0000    11
31.0000    11
48.0000    11
25.0000    11
29.0000    10
34.0000    10
50.0000    10
47.0000    10
39.0000     9
27.0000     9
54.0000     9
33.0000     8
49.0000     8
38.0000     8
19.0000     8
40.0000     8
55.0000     8
32.0000     7
60.0000     7
17.0000     7
26.0000     7
44.0000     6
52.0000     6
37.0000     6
58.0000     6
64.0000     5
46.0000     5
57.0000     5
51.0000     5
16.0000     4
53.0000     4
56.0000     4
43.0000     4
62.0000     4
61.0000     4
1.0000      3
8.0000      3
63.0000     3
41.0000     3
15.0000     2
71.0000     2
20.0000     2
65.0000     2
4.0000      2
6.0000      2
2.0000      1
0.9167      1
14.0000     1
11.0000     1
76.0000     1
59.0000     1
80.0000     1
28.5000     1
32.5000     1
70.0000     1
45.5000     1
13.0000     1
67.0000     1
12.0000     1
0.8333      1
0.6667      1
7.0000      1
3.0000      1
Name: count, dtype: int64

In [ ]:
total_rows = len(data)
total_rows

499

In [ ]:
# Calculate and display the percentage of missing values in each column one by one
for column in data.columns:
  missing_count = data[column].isna().sum()
  missing_percentage = (missing_count / total_rows) * 100
  print(f"Column '{column}' has {missing_count} missing values ({missing_percentage:.2f}%)")

Column 'survived' has 0 missing values (0.00%)
Column 'name' has 0 missing values (0.00%)
Column 'sex' has 0 missing values (0.00%)
Column 'age' has 49 missing values (9.82%)


The percentage of missing values below 20% so we handle numerically with median, categorical with mode. But the categorical data type does not have missing values.

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 499 entries, 0 to 499
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   survived  499 non-null    int64  
 1   name      499 non-null    object 
 2   sex       499 non-null    object 
 3   age       450 non-null    float64
dtypes: float64(1), int64(1), object(2)
memory usage: 19.5+ KB


In [ ]:
data['age'].median()

35.0

In [ ]:
data.isna().sum()

survived    0
name        0
sex         0
age         0
dtype: int64

Success removing NULL values!